### PR TITLE
Change LogDet reduction pattern.

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
The code before change was fine. But would like to avoid CUDAfill_n and atomic.
CUDAfill_n calls cudaMemset which uses the default stream and blocks other streams.
CUDAfill_n was implemented to prevent getting NaN when containers allocates memory. So better to avoid using it outside containers.
In addition, complexDetLog is a very light compute, it is better to have multiple iterations instead of employing more thread blocks.

## What type(s) of changes does this code introduce?
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu [hunk 1]
```diff
@@ -66,35 +65,38 @@ __global__ void computeLogDet_kernel(const int n,
...
-  logdet_vals[threadIdx.x].x = 0.0;
-  logdet_vals[threadIdx.x].y = 0.0;
-  T lu_diag;
-  int n_index = threadIdx.x + block_num * COLBS;
-  if (n_index < n)
+  const int tid      = threadIdx.x;
+  logdet_vals[tid].x = 0.0;
+  logdet_vals[tid].y = 0.0;
+
+  const int num_col_blocks = (n + COLBS - 1) / COLBS;
```
Comment (PDoakORNL) [inline src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu:76]: This needs something like
```c++
static_assert(COLBS % 2 == 0);
```
And a warning for the unwary that this reduction stratedgy only works for powers of 2.  Also do you have profiling information of how much this is worth?  While I too believe it should be faster that's just speculation.

File: src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu [hunk 2]
```diff
@@ -66,35 +65,38 @@ __global__ void computeLogDet_kernel(const int n,
...
+    __syncthreads();
+    if (tid < iend)
     {
-      block_sum_log_det.x += logdet_vals[iv].x;
-      block_sum_log_det.y += logdet_vals[iv].y;
+      logdet_vals[tid].x += logdet_vals[tid + iend].x;
+      logdet_vals[tid].y += logdet_vals[tid + iend].y;
     }
-    // No atomicAdd in cuda 10 for cuComplex/cuDoubleComplex
-    atomicAdd((double*)&(logdets[iw].x), block_sum_log_det.x);
```
Comment (PDoakORNL) [inline src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu:95]: I agree it best to remove the atomic add.

File: src/Platforms/CUDA/CUDAfill.hpp [hunk 3]
```diff
@@ -19,6 +19,14 @@
 
 namespace qmcplusplus
 {
+/** fill device memory with a given value.
```
Comment (PDoakORNL) [inline src/Platforms/CUDA/CUDAfill.hpp:22]: I also dislike cudafill it should be async.

File: src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu:77]: I added a static assert to check exactly power of 2. I have not not looked at the performance but it should be anyway cheaper than getrf.

**Review by PDoakORNL — APPROVED** [review]
I did the benchmarking myself.  Seems to make no difference at all but I think dropping the defensive CUDAfill_n and possibly freeing up some compute on the GPU is worth it.

**Comment by ye-luo** [comment]
Start testing in-house

**Pull request merged**